### PR TITLE
Stateless witness prefetcher changes

### PR DESCRIPTION
**Pull request by karalabe**
Superseeds https://github.com/ethereum/go-ethereum/pull/29035 because OP didn't permit modifications from maintainers...

### CODE REVIEW

**Review by rjl493456442 — APPROVED** [review]

File: core/state/trie_prefetcher.go
Comment (karalabe) [inline core/state/trie_prefetcher.go:306]: Changing the comment is nice, but the code doesn't reflect it :P 

The code should check if `sf.tasks` is nil or not and if not, should keep looping until it becomes so, otherwise we run the risk of receiving a last task and immediately closing down; the close being executed first (remember, select branch evaluation is non-deterministic if multiple channels are ready).

File: core/state/trie_prefetcher.go
Comment (karalabe) [inline core/state/trie_prefetcher.go:149]: These are IMO not good changes. It makes things harder to reason about and the close becomes this magic thing that nukes the prefercher offline, but I'm not sure that's the intended case, since close is also teh thing that *waits* for data to be finished. So we need to figure out what close does: kill it, or wait on it.

File: core/state/trie_prefetcher.go
Comment (karalabe) [inline core/state/trie_prefetcher.go:243]: I don't really see the point of this change, it makes peek useless after close, but close it teh thing that *waits* for all the data to be loaded, so it's kind of ... weird

File: core/state/trie_prefetcher.go
Comment (karalabe) [inline core/state/trie_prefetcher.go:247]: But it can be called multiple times. Close might also not be the best name since we're waiting for it to finish but should AFAIK not kill the thing.

File: core/state/trie_prefetcher.go
Comment (karalabe) [inline core/state/trie_prefetcher.go:301]: I'm unsure about this code path here with the rewrite. Do we want to allow retrieval from a live prefetcher? If yes, why? Perhaps for tx boundaries? We should really document it somewhere why - if - it's needed. It's a very specific use case.

**Comment by karalabe** [comment]
As I understand the difference between the old and new prefetcher is (should be) as follows:

- Old pre-fetcher:
  - Purpose is to warm up the trie **during** execution, so that while we're crunching some EVM code, our disk is kept busy pulling in data that the hasher will need at the end.
  - All operations are async, running in the background, the most important thing is to never ever block. If we get more useful data great, if less, thats life, but we should never hold up execution.
  - When execution reaches a boundary (IntermediateRoot pre-Byzantium; or Finalize after Byzantium), insta-terminate all pre-fetchers to avoid the main committer thread from racing for disk accesses. Whatever we managed to load will be used, the rest pulled on demand.

- New pre-fetcher:
  - Purpose is to act as a witness constructor (write only for now) **during** execution, so that while we're crunching some EVM code, our disk is kept busy pulling in data that both the hasher hasher, but also a cross-validator will need at the end.
  - **Almost** all operations are async, running in the background, the most important thing is to never ever block **during** EVM execution. However, on commit boundaries we have to switch to blocking mode, since the witness needs all data, not just whatever we loaded until that point in time.
  - When execution reaches a boundary (IntermediateRoot pre-Byzantium; or Finalize after Byzantium), **wait** for all pre-fetchers to finish. This will **block** the main committer thread, but ideally if we're not loading junk, it should be all the same, the data needs to be loaded anyway to commit. For the witness, the data **must** be loaded, before tries are mutated.

- (Threading) Quirks:
  - Pre-Byzantium does an IntermediateRoot call between each transaction. A witness pre-fetcher for that block range *must* support stopping after a transaction, collecting the witness; then continuing against the next transaction, collecting witnesses from updated tries. This is significantly more complex from both a witness and a threading perspective, to have data across tries. Given that pre-Byzantium is ancient, it doesn't make sense to support it, but we need to very explicitly handle / reject that case, otherwise it's going to be "weird" trying to understand the code.
  - The old pre-fetcher was best-effort, with no guarantees on correctness (as to how much and what data it loaded). The new pre-fetcher needs to be correct to construct a proper witness, so sometimes blocking is necessary. That however means that code paths need to be re-thought, as we still want to maximise the main EVM execution pathways even whilst waiting for data. Particularly, when terminating a pre-fetcher (i.e. waiting), we should start integrating results from finished subfetchers before waiting for all storage tries to finish loading.

- Qustions:
  - Does slot mutation order make the witness different? I.e. If i change 3 slots in a contract (including delete/create), does the order of applying them change what trie nodes we need? Because if so, there might be a hidden step still needed during commit to add prefetch tasks (?)

File: core/state/trie_prefetcher.go
Comment (jwasinger) [reply core/state/trie_prefetcher.go:301]: When we call `updateTrie` on a state object, we attempt to source the trie from the prefetcher.  So copying from a live prefetcher is used here to preserve that functionality.

**Comment by jwasinger** [comment]
> Purpose is to act as a witness constructor (write only for now) during execution, so that while we're crunching some EVM code, our disk is kept busy pulling in data that both the hasher hasher, but also a cross-validator will need at the end.

It's actually meant to gather witnesses for read values.  In the stateless witness builder [PR](https://github.com/ethereum/go-ethereum/pull/29029), I gather write witnesses from committing the tries.

But iirc, earlier on the call today you mentioned not tying the retrieval of write witnesses to the commit operation, which would change the assumptions from my original changes.

**Comment by karalabe** [comment]
Screenshot 2024-05-07 at 09 22 54

File: core/state/state_object.go
Comment (karalabe) [inline core/state/state_object.go:138]: FWIW, `skipPrefetcher` is kind of an ugly hack, I just wanted to avoid the lack-of-snapshot poking into the prefetcher. Open to cleaner suggestions.

File: core/state/state_object.go
Comment (rjl493456442) [inline core/state/state_object.go:219]: It's probably problematic?

- Prefetcher is enabled if the associated snapshot layer is available
- Retrievals from snapshot can succeed if the requested item is covered 
- Retrievals from snapshot can fail if the requested item is not covered yet(under the generation)
- The scenario is possible a few data retrieval through snapshot succeed and some prefetching tasks have been scheduled; then a retrieval fails and the prefetch'd trie is abandoned

The witness of the trie could be incomplete

Comment (jwasinger) [reply core/state/state_object.go:219]: The parameter to `getTrie` determines whether or not it will attempt to load the trie from the prefetcher, so it will not stop the prefetcher prematurely here.

Afaict the changes that need to be made for witness-gathering-enabled path are:
* if snap is enabled, prefetch storage access lists on each call to `GetCommittedState` ([i.e.](https://github.com/ethereum/go-ethereum/pull/29719/files#diff-493fcc97ac5eefa37eef088731c482655afbae6c1ebaead72a88b7746897c606R215-R220))
* terminate the prefetcher synchronously at the start of the `IntermediateRoot` ([i.e.](https://github.com/ethereum/go-ethereum/pull/29719/files#diff-c3757dc9e9d868f63bc84a0cc67159c1d5c22cc5d8c9468757098f0492e0658cR896))
* later in `IntermediateRoot`: accumulate storage access lists from the trie of each mutated state object ([i.e.](https://github.com/ethereum/go-ethereum/pull/29719/files#diff-c3757dc9e9d868f63bc84a0cc67159c1d5c22cc5d8c9468757098f0492e0658cR1317))
* in statedb commit, iterate set of non-mutated state objects, if we can resolve a trie from the prefetcher, add its access list to the witness ([i.e.](https://github.com/ethereum/go-ethereum/pull/29719/files#diff-c3757dc9e9d868f63bc84a0cc67159c1d5c22cc5d8c9468757098f0492e0658cR1401)).

Comment (rjl493456442) [reply core/state/state_object.go:219]: > it will not stop the prefetcher prematurely here.

yes, it won't abort the prefetching, but the issue is the `trie` will be resolved from database and be linked to stateObject. The trie in prefetcher will never be retrieved later on.

Comment (jwasinger) [reply core/state/state_object.go:219]: Ah I see.  Yeah, it's an issue.

Comment (karalabe) [reply core/state/state_object.go:219]: It does work as intended, but I agree that the intent was wrong :)

Comment (karalabe) [reply core/state/state_object.go]: I pushed a change to fix this. Whilst the optimization of loading *from the EVM excution into* the prefetcher is still lost (cleaner IMO); the new code forcefully overwrites the cached trie from the prefetcher on update if available. To make it less magical, I made this overwrite more explicit in the update code instead of relying on the cached trie magically getting swapped out based on some weird call sequence.

File: core/state/state_object.go
Comment (rjl493456442) [inline core/state/state_object.go:159]: typo `pretecher` => `prefetcher`

File: core/state/state_object.go
Comment (rjl493456442) [inline core/state/state_object.go:293]: It could be problematic.

In `updateRoot`, it will:
- flush the pending storage changes into the hasher(trie)
- compute the new storage root

at that time, prefetcher is already terminated. Unfortunately, it might be possible that there are still some cached storage changes, e.g. caused by some rules post-execution (e.g. withdrawal, although withdrawal won't touch storages, but we might have more post-execution rules in the future)

Comment (rjl493456442) [reply core/state/state_object.go:293]: EDIT: all the `finalization` has been done in `IntermediateRoot` before the prefetcher termination. So the issue is not relevant anymore.


But we should remove `	s.finalise` in `stateObject.updateTrie` later as it's useless.

File: core/state/trie_prefetcher.go
Comment (AungThuSoe24) [inline core/state/trie_prefetcher.go:19]: BTC 2

**Pull request merged**
